src/proxy/canonicalize.rs

```diff
@@ -167,30 +178,41 @@ impl Future for Task {
                             // service with it and set a delay that will notify
                             // when the resolver should be consulted again.
                             let resolved = NameAddr::new(refine.name, self.original.port());
-                            if self.resolved.as_ref() != Some(&resolved) {
+                            if self.resolved.get() != Some(&resolved) {
                                 let err = self.tx.try_send(resolved.clone()).err();
                                 if err.map(|e| e.is_disconnected()).unwrap_or(false) {
                                     return Ok(().into());
                                 }
 
-                                self.resolved = Some(resolved);
+                                self.resolved = Cache::Resolved(resolved);
                             }
 
                             State::ValidUntil(Delay::new(refine.valid_until))
                         }
                         Err(e) => {
-                            if self.resolved.is_none() {
-                                error!("failed to refine {}: {}", self.original.name(), e);
+                            if self.resolved == Cache::AwaitingInitial {
+                                // The service needs a value, so we need to
+                                // publish the original name so it can proceed.
+                                warn!(
+                                    "failed to refine {}: {}; using original name",
+                                    self.original.name(),
+                                    e,
+                                );
                                 let err = self.tx.try_send(self.original.clone()).err();
                                 if err.map(|e| e.is_disconnected()).unwrap_or(false) {
                                     return Ok(().into());
                                 }
 
-                                // Pretend the original name was resolved so
-                                // that we don't re-publish on subsequent errors.
-                                self.resolved = Some(self.original.clone());
+                                // There's now no need to re-publish the
+                                // original name on subsequent failures.
+                                self.resolved = Cache::Unresolved;
                             } else {
-                                warn!("failed to refine {}: {}, falling back", self.original.name(), e);
+                                debug!(
+                                    "failed to refresh {}: {}; cache={:?}",
```

Comment:
:+1: for "failed to refresh", that definitely makes it clearer that we were just trying to update a previously resolved name.